### DISCUSSION TITLE
Implementing viewContainers and treeViews

### DISCUSSION THREAD

**Discussion by castrob** [CodinGame/monaco-vscode-api#495] [Q&A] [answered]
Hello!

I'm working on implementing a custom `treeView` within my project, inside a custom `viewContainer`, to be listed alongside other containers like Explorer and Extensions.

I've been exploring different approaches by looking at VSCode and the demo, and I've come across two main ways of doing this. I would appreciate guidance on which approach is recommended:

1. Implementing via extensions (registerExtension, getApi()...), using the VSCode docs, such as [this one](https://code.visualstudio.com/api/extension-guides/tree-view). Is this possible? I've seen on the demo an example of a customView using HTML implementation, but not anything like treeView itself.

2. Implementing using the "VSCode contribution" method, similar to the `UserDataSyncWorkbenchContribution -> registerViews -> ... -> UserDataSyncDataViews.registerMachinesView` approach.

When attempting the second approach, I've had difficulty importing certain types, such as `TreeView`, `Registry` (from `vs/platform/registry`), etc. I understand that this approach is more "internal" than the first, but is it even possible? Could I be misunderstanding how to extend this project and VSCode for web?

Thanks for your help!

**Comment by CGNonofr** [comment, marked as answer]
Hello!

> Implementing via extensions (registerExtension, getApi()...), using the VSCode docs, such as [this one](https://code.visualstudio.com/api/extension-guides/tree-view). Is this possible? I've seen on the demo an example of a customView using HTML implementation, but not anything like treeView itself.

There's no reason for it to be not possible

> Implementing using the "VSCode contribution" method, similar to the UserDataSyncWorkbenchContribution -> registerViews -> ... -> UserDataSyncDataViews.registerMachinesView approach.

It's more internal indeed, but it may allow more customizations

Regardless of what you are trying to achieve, If what you are trying to achieve is doable via the VSCode extension api, I would recommend doing it that way because:
- The api is well documented and stabilized
- It wouldn't rely on internal types that may be renamed, moved or refactored
- Not everything is properly exported
- There are some issues with using internal VSCode structures, like the usage of const enums that forces you to disable typescript module isolation
- You are able to use the official VSCode extension building tools

**Reply by castrob** [reply]
Thanks for the advice! I was able to do it using the extension way, it was kinda difficult to handle file paths and to understand the registerFileURL function, but I managed it to work!
